### PR TITLE
Consent hiu workflows

### LINKED ISSUE #12: Consent Workflow on HRP
As a developer, I would like to build the workflows required for consent request which needs to be sent to Gateway. And once is created at Gateway, it should be stored in the database so that next time when its enquired, it can be returned from database itself.

### LINKED ISSUE #46: Error Handling - Add more constructs to handle errors better.
- Have a http_status.valueof(response).is2xxSuccessful check everywhere.
- Handle Gateway error response by propagating the same http error code
- Service layer can send plain pojos as response and controller can wrap into ResponseEntity

### PR DESCRIPTION
**Pull request by atulai-sg**
Added logical pieces for Consent Management

Fixes #12 #13 #14 #46 #45

### CODE REVIEW

**Review by chandrashekar-s — COMMENTED** [review]
Thanks for the changes @atulai-sg, they look good. There are only few minor comments, please have a look.

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java [hunk 1]
```diff
@@ -334,24 +338,60 @@ public void setHipOnAddCareContextResponse(
...
+    requestLog.setGatewayRequestId(initConsentRequest.getRequestId());
+    requestLog.setStatus(status);
+    if (StringUtils.isNotBlank(error)) {
+      requestLog.setError(error);
+    }
+    mongoTemplate.save(requestLog);
+  }
+
+  @Transactional
+  public void updateConsentResponse(String requestId, String consentRequestId) {
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:372]: In the first method, we dont have entire consent context, we do have just consentRequestId.

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/PatientService.java [hunk 2]
```diff
@@ -164,15 +165,26 @@ public void addPatientCareContexts(LinkRecordsRequest linkRecordsRequest) {
     log.info("Successfully Added Patient careContexts");
   }
 
-  public void addConsentArtefacts(HIPNotifyRequest hipNotifyRequest) {
-    String abhaAddress =
-        hipNotifyRequest.getHIPNotification().getConsentDetail().getPatient().getId();
+  public void addConsent(String abhaAddress, Consent consent) throws IllegalDataStateException {
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/PatientService.java:168]: Cool

File: src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java [hunk 3]
```diff
@@ -1,36 +1,142 @@
...
+        notifyHIURequest.getRequestId(),
+        "Something went wrong while executing consent on status",
+        RequestStatus.CONSENT_HIU_NOTIFY_ERROR);
   }
 
   @Override
-  public ResponseEntity<GatewayCallbackResponse> consentOnFetch(OnFetchRequest onFetchRequest) {
-    // TODO: Implement consent on fetch workflow logic.
-    return null;
+  public void consentOnFetch(OnFetchRequest onFetchRequest) throws IllegalDataStateException {
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hiu/hrp/consent/GatewayCallbackService.java:116]: Thanks for the suggestion, but setting Transation annotation alone is not sufficient. We need to enable the transaction management. [Reference](https://www.geeksforgeeks.org/spring-boot-transaction-management-using-transactional-annotation/). Made changes.

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/PatientService.java [hunk 4]
```diff
@@ -164,15 +165,26 @@ public void addPatientCareContexts(LinkRecordsRequest linkRecordsRequest) {
...
+    List<Consent> consents = patient.getConsents();
+    for (Consent storedConsent : consents) {
+      if (storedConsent
+          .getConsentDetail()
+          .getConsentId()
+          .equals(consent.getConsentDetail().getConsentId())) {
+        String message =
+            String.format(
+                "Consent %s already exists for patient %s: ", consent.toString(), abhaAddress);
+        log.info(message);
```
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/PatientService.java:182]: ack

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (chandrashekar-s) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:359]: Agreed. The only concern I had was the `FieldIdentifiers.RESPONSE_DETAILS` field is taking different structures in each method, I was trying to see if we can get a uniform structure. If this decoding can be handled properly while reading the data based on `STATUS` then its fine.

**Review by chandrashekar-s — APPROVED** [review]
Thanks @atulai-sg for addressing review comments.

File: src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (atulai-sg) [reply src/main/java/com/nha/abdm/wrapper/hip/hrp/database/mongo/services/RequestLogService.java:359]: cool

**Pull request merged**